### PR TITLE
BTFS-1573

### PR DESCRIPTION
**Pull request by steveyeom**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior?** (You can also refer to a JIRA ticket here)

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **What dependencies / modules need to be updated as pre-requisites?** (Include relevant PR links here)

* **Description of changes**

---

* **Please check if the PR fulfills these requirements**
- [ ] The subject of this PR contains a JIRA ticket BTFS-xxxx (N/A for community PRs)
- [ ] PR description and commit messages are [good and meaningful](https://chris.beams.io/posts/git-commit/)
- [ ] Unit tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

* **Please make sure the following procedures have been applied before requesting reviewers**
- [ ] Code changes closely follow [Go Code Review Comments](https://github.com/golang/go/wiki/CodeReviewComments)
- [ ] Code has been *rebased* against latest master (`git merge` not recommended, unless you know what you are doing)
- [ ] Code changes have run through `go fmt`
- [ ] Code changes have run through `go mod tidy`
- [ ] All unit tests passed locally (`make test_go_test`)

### CODE REVIEW

**Comment by steveyeom** [comment]
Tested in my AWS linux VM with and without the PR. It is verified in that linux.

File: core/coreunix/add.go [hunk 1]
```diff
@@ -365,6 +367,7 @@ func (adder *Adder) AddAllAndPin(file files.Node) (ipld.Node, error) {
 		adder.unlocker = adder.gcLocker.PinLock()
 	}
 	defer func() {
+		adder.AddInProcess = false
```
Comment (taiyangc) [inline core/coreunix/add.go]: This is never set to true? Then I wonder why you need it.

File: core/coreunix/test/metadata_test.go [hunk 2]
```diff
@@ -153,6 +153,7 @@ func addFileToBtfs(node *core.IpfsNode, data []byte, metadata string) (ipath.Pat
 	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
 	defer cancel()
 
+	adder.AddInProcess = true
```
Comment (taiyangc) [inline core/coreunix/test/metadata_test.go]: This is weird - why would a test force change this variable to true? This variable should be set at the beginning of the actual processing code.

File: core/coreunix/metadata.go [hunk 3]
```diff
@@ -6,12 +6,10 @@ import (
 	"encoding/json"
 	"errors"
 	"fmt"
-	gopath "path"
 
+	chunker "github.com/TRON-US/go-btfs-chunker"
 	core "github.com/TRON-US/go-btfs/core"
 	"github.com/TRON-US/go-btfs/pin"
-
-	chunker "github.com/TRON-US/go-btfs-chunker"
```
Comment (taiyangc) [inline core/coreunix/metadata.go]: original groups are correct

File: core/coreunix/test/metadata_test.go [hunk 4]
```diff
@@ -173,6 +174,20 @@ func addFileToBtfs(node *core.IpfsNode, data []byte, metadata string) (ipath.Pat
 		return nil, ctx.Err()
 	}
 
+	// The following adder.AddInProcess-for-loop mechanism is to wait for
+	// AddAllAndPin() to finish its callng of adder.PinRoot() since AddAllAndPin()
```
Comment (taiyangc) [inline core/coreunix/test/metadata_test.go]: The assumption that `AddAllAndPin` can return without finish calling `adder.PinRoot` is already a wrong assumption. `AddAllAndPin` should be a synchronous process where it should only return when all the sub-logic is done. So adding a workaround to wait for an async routine to finish here is a hack that should not be used.

Comment (steveyeom) [reply core/coreunix/test/metadata_test.go]: Checking the code. But there could be some reasons to do that. Without the metata tests, the current logic of AddAllAndPin works.

Comment (steveyeom) [reply core/coreunix/test/metadata_test.go]: Like progressReader, i don't have an option to check details.

Comment (steveyeom) [reply core/coreunix/test/metadata_test.go]: For the case of directory, AddAllAndPin() updates progressReader by sending file root node to the adder.Out channel. Then at the last line of the function, PinRoot() is to called to add the top node of the directory.

**Comment by steveyeom** [comment]
Updated the PR to be simpler. I am testing the updated PR in AWS linux. Without a failure for 6 times.

File: core/coreunix/metadata.go (hunk 3 above)
Comment (steveyeom) [reply core/coreunix/metadata.go]: Fixed.

**Pull request merged**